src/finch/tensor.py

```diff
@@ -785,7 +785,25 @@ def _reduce(x: Tensor, fn: Callable, axis, dtype=None):
         result = fn(x._obj, dims=axis)
     else:
         result = fn(x._obj)
-    return Tensor(result)
+
+    if np.isscalar(result):
+        result = jl.Tensor(
+            jl.Element(
+                jc.convert(x.dtype, 0),
+                np.array(result, dtype=jl_dtypes.jl_to_np_dtype[x.dtype])
+            )
+        )
+
+    result = Tensor(result)
+    if dtype != "skip":
+        # dtype casting rules for `sum` and `prod`
+        if dtype is not None:
+            result = astype(result, dtype)
+        elif jl.seval(f"{x.dtype} <: Unsigned"):
+            result = astype(result, jl_dtypes.uint)
+        elif jl.seval(f"{x.dtype} <: Signed"):
+            result = astype(result, jl_dtypes.int_)
```

Comment:
The dtype promotion rules are invalid in the 2023 version of the Array API standard -- we shouldn't write code to cater to an old version of the standard.

Reply:
Ah, that's my bad, it does exist in 2023.

Reply:
I think they are still valid in `2023` version: https://data-apis.org/array-api/2023.12/API_specification/generated/array_api.sum.html

> if x has a signed integer data type (e.g., int16), the returned array must have the default integer data type.

> if x has an unsigned integer data type (e.g., uint16), the returned array must have an unsigned integer data type having the same number of bits as the default integer data type (e.g., if the default integer data type is int32, the returned array must have a uint32 data type).